compiler/AST/wellknown.cpp

```diff
@@ -0,0 +1,270 @@
+
+      delete dtString;
+
+      dtString = NULL;
+    }
+  }
+}
+
+struct WellKnownFn
+{
+  const char* name;
+  FnSymbol**  fn;
+  Flag        flag;
+  FnSymbol*   lastNameMatchedFn;
+};
+
+// These functions are a required part of the compiler/module interface.
+// They should generally be marked export so that they are always
+// resolved.
+static WellKnownFn sWellKnownFns[] = {
+  {
+    "chpl_here_alloc",
+    &gChplHereAlloc,
+    FLAG_LOCALE_MODEL_ALLOC
+  },
+
+  {
+    "chpl_here_free",
+    &gChplHereFree,
+    FLAG_LOCALE_MODEL_FREE
+  },
+
+  {
+    "chpl_doDirectExecuteOn",
+    &gChplDoDirectExecuteOn,
+    FLAG_UNKNOWN
+  },
+
+  {
+    "_build_tuple",
+    &gBuildTupleType,
+    FLAG_BUILD_TUPLE_TYPE
+  },
+
+  {
+    "_build_tuple_noref",
+    &gBuildTupleTypeNoRef,
+    FLAG_BUILD_TUPLE_TYPE
+  },
+
+  {
+    "*",
+    &gBuildStarTupleType,
+    FLAG_BUILD_TUPLE_TYPE
+  },
+
+  {
+    "_build_star_tuple_noref",
+    &gBuildStarTupleTypeNoRef,
+    FLAG_BUILD_TUPLE_TYPE
+  },
+
+  {
+    "chpl_delete_error",
+    &gChplDeleteError,
+    FLAG_UNKNOWN
+  },
+
+  {
+    "printModuleInit",
+    &gPrintModuleInitFn,
+    FLAG_PRINT_MODULE_INIT_FN
+  }
+};
+
+void gatherWellKnownFns() {
+  int nEntries = sizeof(sWellKnownFns) / sizeof(sWellKnownFns[0]);
+
+  // Harvest well-known functions from among the global fn definitions.
+  // We check before assigning to the associated global to ensure that it
+  // is null.  In that way we can flag duplicate definitions.
+  forv_Vec(FnSymbol, fn, gFnSymbols) {
+    for (int i = 0; i < nEntries; ++i) {
+      WellKnownFn& wkfn = sWellKnownFns[i];
+
+      if (strcmp(fn->name, wkfn.name) == 0) {
+        wkfn.lastNameMatchedFn = fn;
+
+        if (wkfn.flag == FLAG_UNKNOWN || fn->hasFlag(wkfn.flag) == true) {
+          if (*wkfn.fn != NULL) {
+            USR_WARN(fn,
+                     "'%s' defined more than once in Chapel internal modules.",
+                     wkfn.name);
+          }
+
+          *wkfn.fn = fn;
+        }
+      }
+    }
+  }
+
+  if (fMinimalModules == false) {
+    for (int i = 0; i < nEntries; ++i) {
+      WellKnownFn& wkfn        = sWellKnownFns[i];
+      FnSymbol*    lastMatched = wkfn.lastNameMatchedFn;
+      FnSymbol*    fn          = *wkfn.fn;
+
+      if (lastMatched == NULL) {
+        USR_FATAL_CONT("Function '%s' must be defined in the "
+                       "Chapel internal modules.",
+                       wkfn.name);
+
+      } else if (fn == NULL) {
+        USR_FATAL_CONT(fn,
+                       "The '%s' function is missing a required flag.",
+                       wkfn.name);
+      }
+    }
+
+    USR_STOP();
+  }
+}
+
+std::vector<FnSymbol*> getWellKnownFunctions()
+{
+  std::vector<FnSymbol*> fns;
+
+  int nEntries = sizeof(sWellKnownFns) / sizeof(sWellKnownFns[0]);
+
+  for (int i = 0; i < nEntries; ++i) {
+    WellKnownFn& wkfn = sWellKnownFns[i];
+    fns.push_back(*wkfn.fn);
+  }
+
```

Comment:
I am sure your code is right but I can't see why gPrintModuleInitFn wasn't pushed on fns by the general
loop over sWellKnownFns.